### PR TITLE
Add an option to configure a timeout for plugins

### LINKED ISSUE #100: Trollflow2 stops processing if plugin gets stuck (messages go unprocessed, no error message in log)
**Describe the bug**

I configured trollflow2 to call the plugin `trollflow2.plugins.covers`.  When this one gets stuck, which in this case happens due to the pyresample bug https://github.com/pytroll/pyresample/issues/329, trollflow2 does not recover but stops processing forever. 
The cause is not clear from log messages and it took me substantial detective work with [pyrasite](https://github.com/lmacken/pyrasite) to identify the root cause in the first place. For example, in our test operational chain, from 2021-02-07 at 12:58 UTC, in a test production chain using latest trollflow2 master (more specifically the branch corresponding to #97), trollflow2 was no longer processing incoming messages.  There are no warning or error messages that may explain this.  At 12:55:03, two messages were received and processed.  At 12:58:05, two messages were received and were not processed, nor were any subsequent messages until I restarted trollflow2 the following day.  Since restart, messages have been processed again.

```
[DEBUG: 2021-02-07 12:55:03 : Listener] New message received: pytroll://collection/metop/avhrr/eumetcast collection pytroll@oflks333.dwd.de 2021-02-07T12:55:03.001058 v1.01 application/json {"path": "", "start_time": "2021-02-07T11:07:03", "end_time": "2021-02-07T11:16:03", "processing_time": "2021-02-07T11:39:48", "sensor": "avhrr/3", "platform_name": "Metop-A", "collection_area_id": "nparktis1km", "collection": [{"start_time": "2021-02-07T11:07:03", "end_time": "2021-02-07T11:10:03", "uri": "/data/pytroll/IN/EMC_AVR/AVHR_xxx_1B_M02_20210207110703Z_20210207111003Z_N_O_20210207113948Z", "uid": "AVHR_xxx_1B_M02_20210207110703Z_20210207111003Z_N_O_20210207113948Z"}, {"start_time": "2021-02-07T11:10:03", "end_time": "2021-02-07T11:13:03", "uri": "/data/pytroll/IN/EMC_AVR/AVHR_xxx_1B_M02_20210207111003Z_20210207111303Z_N_O_20210207114013Z", "uid": "AVHR_xxx_1B_M02_20210207111003Z_20210207111303Z_N_O_20210207114013Z"}, {"start_time": "2021-02-07T11:13:03", "end_time": "2021-02-07T11:16:03", "uri": "/data/pytroll/IN/EMC_AVR/AVHR_xxx_1B_M02_20210207111303Z_20210207111603Z_N_O_20210207114037Z", "uid": "AVHR_xxx_1B_M02_20210207111303Z_20210207111603Z_N_O_20210207114037Z"}]}
[DEBUG: 2021-02-07 12:55:03 : Listener] New message received: pytroll://collection/metop/avhrr/eumetcast collection pytroll@oflks333.dwd.de 2021-02-07T12:55:03.002003 v1.01 application/json {"path": "", "start_time": "2021-02-07T11:13:03", "end_time": "2021-02-07T11:16:03", "processing_time": "2021-02-07T11:40:37", "sensor": "avhrr/3", "platform_name": "Metop-A", "collection_area_id": "npgroenland2200m", "collection": [{"start_time": "2021-02-07T11:13:03", "end_time": "2021-02-07T11:16:03", "uri": "/data/pytroll/IN/EMC_AVR/AVHR_xxx_1B_M02_20210207111303Z_20210207111603Z_N_O_20210207114037Z", "uid": "AVHR_xxx_1B_M02_20210207111303Z_20210207111603Z_N_O_20210207114037Z"}]}
[INFO: 2021-02-07 12:55:03 : trollflow2.launcher] Distributed processing not configured, using default scheduler
[INFO: 2021-02-07 12:55:03 : trollflow2.plugins] Generating scene
(...)
[DEBUG: 2021-02-07 12:55:48 : trollflow2.launcher] Cleaning up
[INFO: 2021-02-07 12:55:49 : trollflow2.launcher] All files produced nominally in 0:00:46.061789.
[INFO: 2021-02-07 12:55:49 : trollflow2.launcher] Distributed processing not configured, using default scheduler
[INFO: 2021-02-07 12:55:49 : trollflow2.plugins] Generating scene
(...)
[DEBUG: 2021-02-07 12:56:25 : trollflow2.launcher] Cleaning up
[INFO: 2021-02-07 12:56:25 : trollflow2.launcher] All files produced nominally in 0:00:36.357745.
[DEBUG: 2021-02-07 12:58:05 : Listener] New message received: pytroll://collection/metop/avhrr/eumetcast collection pytroll@oflks333.dwd.de 2021-02-07T12:58:05.614928 v1.01 application/json {"path": "", "start_time": "2021-02-07T11:16:03", "end_time": "2021-02-07T11:19:03", "processing_time": "2021-02-07T12:55:34", "sensor": "avhrr/3", "platform_name": "Metop-A", "collection_area_id": "nparktis1km", "collection": [{"start_time": "2021-02-07T11:16:03", "end_time": "2021-02-07T11:19:03", "uri": "/data/pytroll/IN/EMC_AVR/AVHR_xxx_1B_M02_20210207111603Z_20210207111903Z_N_O_20210207125534Z", "uid": "AVHR_xxx_1B_M02_20210207111603Z_20210207111903Z_N_O_20210207125534Z"}]}
[DEBUG: 2021-02-07 12:58:05 : Listener] New message received: pytroll://collection/metop/avhrr/eumetcast collection pytroll@oflks333.dwd.de 2021-02-07T12:58:05.615918 v1.01 application/json {"path": "", "start_time": "2021-02-07T11:16:03", "end_time": "2021-02-07T11:19:03", "processing_time": "2021-02-07T12:55:34", "sensor": "avhrr/3", "platform_name": "Metop-A", "collection_area_id": "npgroenland2200m", "collection": [{"start_time": "2021-02-07T11:16:03", "end_time": "2021-02-07T11:19:03", "uri": "/data/pytroll/IN/EMC_AVR/AVHR_xxx_1B_M02_20210207111603Z_20210207111903Z_N_O_20210207125534Z", "uid": "AVHR_xxx_1B_M02_20210207111603Z_20210207111903Z_N_O_20210207125534Z"}]}
[DEBUG: 2021-02-07 12:58:43 : Listener] New message received: pytroll://collection/metop/avhrr/eumetcast collection pytroll@oflks333.dwd.de 2021-02-07T12:58:43.487091 v1.01 application/json {"path": "", "start_time": "2021-02-07T11:19:03", "end_time": "2021-02-07T11:22:03", "processing_time": "2021-02-07T12:56:47", "sensor": "avhrr/3", "platform_name": "Metop-A", "collection_area_id": "nparktis1km", "collection": [{"start_time": "2021-02-07T11:19:03", "end_time": "2021-02-07T11:22:03", "uri": "/data/pytroll/IN/EMC_AVR/AVHR_xxx_1B_M02_20210207111903Z_20210207112203Z_N_O_20210207125647Z", "uid": "AVHR_xxx_1B_M02_20210207111903Z_20210207112203Z_N_O_20210207125647Z"}]}
[DEBUG: 2021-02-07 12:58:43 : Listener] New message received: pytroll://collection/metop/avhrr/eumetcast collection pytroll@oflks333.dwd.de 2021-02-07T12:58:43.489299 v1.01 application/json {"path": "", "start_time": "2021-02-07T11:19:03", "end_time": "2021-02-07T11:22:03", "processing_time": "2021-02-07T12:56:47", "sensor": "avhrr/3", "platform_name": "Metop-A", "collection_area_id": "npgroenland2200m", "collection": [{"start_time": "2021-02-07T11:19:03", "end_time": "2021-02-07T11:22:03", "uri": "/data/pytroll/IN/EMC_AVR/AVHR_xxx_1B_M02_20210207111903Z_20210207112203Z_N_O_20210207125647Z", "uid": "AVHR_xxx_1B_M02_20210207111903Z_20210207112203Z_N_O_20210207125647Z"}]}
[DEBUG: 2021-02-07 13:01:03 : Listener] New message received: pytroll://collection/metop/avhrr/eumetcast collection pytroll@oflks333.dwd.de 2021-02-07T13:01:03.001442 v1.01 application/json {"path": "", "start_time": "2021-02-07T11:13:03", "end_time": "2021-02-07T11:22:03", "processing_time": "2021-02-07T11:40:37", "sensor": "avhrr/3", "platform_name": "Metop-A", "collection_area_id": "euro_ma", "collection": [{"start_time": "2021-02-07T11:13:03", "end_time": "2021-02-07T11:16:03", "uri": "/data/pytroll/IN/EMC_AVR/AVHR_xxx_1B_M02_20210207111303Z_20210207111603Z_N_O_20210207114037Z", "uid": "AVHR_xxx_1B_M02_20210207111303Z_20210207111603Z_N_O_20210207114037Z"}, {"start_time": "2021-02-07T11:16:03", "end_time": "2021-02-07T11:19:03", "uri": "/data/pytroll/IN/EMC_AVR/AVHR_xxx_1B_M02_20210207111603Z_20210207111903Z_N_O_20210207125534Z", "uid": "AVHR_xxx_1B_M02_20210207111603Z_20210207111903Z_N_O_20210207125534Z"}, {"start_time": "2021-02-07T11:19:03", "end_time": "2021-02-07T11:22:03", "uri": "/data/pytroll/IN/EMC_AVR/AVHR_xxx_1B_M02_20210207111903Z_20210207112203Z_N_O_20210207125647Z", "uid": "AVHR_xxx_1B_M02_20210207111903Z_20210207112203Z_N_O_20210207125647Z"}]}
```

There was no traceback reported around 12:58.  The most recent traceback in `trollflow2.log` was reported at 10:00:24 and related to channel 3B (`KeyError: "Could not load DataID(name='3b', ...) from any provided files`).  Since this was nearly three hours before processing of incoming messages stopped and I have had this message several times per day on 6 and 7 February, I believe this is unrelated.

I am running all processing via supervisord.  The supervisord logfile does not contain any additional information or anything unusual around the indicated time.  At the time, supervisorctl showed all processes as running with the same uptime.

I had started `trollflow2` with:

```
/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py38/bin/satpy_launcher.py -n localhost /opt/pytroll/pytroll_inst/config/trollflow2.yaml
```

No other Pytroll packages in the chain reported any tracebacks on 2021-08-07.

Apart from the tracebacks reported above, no messages were reported with loglevel ERROR.  There were some messages reported with loglevel WARNING, but I believe those were also unrelated as they were logged every time:

```
[WARNING: 2021-02-07 12:50:55 : satpy.scene] The following datasets were not created and may require resampling to be generated: DataID(name='night_fog'), DataID(name='day_microphysics'), DataID(name='cloudtop'), DataID(name='night_microphysics'), DataID(name='overview'), DataID(name='ir108_3d'), DataID(name='ir_cloud_day')
[WARNING: 2021-02-07 12:50:56 : pyspectral.rsr_reader] Inconsistent instrument/satellite input - instrument set to avhrr/3
[WARNING: 2021-02-07 12:50:56 : pyspectral.rsr_reader] Inconsistent instrument/satellite input - instrument set to avhrr/3
[WARNING: 2021-02-07 12:50:57 : pyspectral.rsr_reader] Inconsistent instrument/satellite input - instrument set to avhrr/3
[WARNING: 2021-02-07 12:55:03 : satpy.readers.yaml_reader] No filenames found for reader: avhrr_l1b_aapp
[WARNING: 2021-02-07 12:55:16 : satpy.scene] The following datasets were not created and may require resampling to be generated: DataID(name='night_fog'), DataID(name='ir108_3d'), DataID(name='cloudtop'), DataID(name='night_microphysics')
[WARNING: 2021-02-07 12:55:49 : satpy.readers.yaml_reader] No filenames found for reader: avhrr_l1b_aapp
[WARNING: 2021-02-07 12:55:55 : satpy.scene] The following datasets were not created and may require resampling to be generated: DataID(name='night_fog'), DataID(name='ir108_3d'), DataID(name='cloudtop'), DataID(name='night_microphysics')
```

**To Reproduce**

The plugin `covers` gets stuck due to https://github.com/pytroll/pyresample/issues/329 .  This MCVE triggers the same, although that's pyresample's fault and not trollflow2's:

```python
from datetime import datetime
from satpy.utils import debug_on; debug_on()
from trollflow2.plugins import get_scene_coverage
get_scene_coverage("Metop-C", datetime(2021, 2, 9, 14, 22, 3),
        datetime(2021, 2, 9, 16, 4, 3), "avhrr/3", "npgroenland2200m")
```

The area `npgroenland2200m` is defined as:

```
npgroenland2200m:
  description: npgroenland2200m
  projection:
    proj: stere
    lon_0: -30.0
    lat_0: 72.0
    lat_ts: 60.0
    ellps: WGS84
  shape:
    height: 1280
    width: 1024
  area_extent:
    lower_left_xy: [-1139231.5127295423, -1463534.2335492456]
    upper_right_xy: [1225459.1694661214, 1487863.414857883]
```

Due to https://github.com/pytroll/pyresample/issues/329 this never completes.

**Expected behavior**

I expect that if there is a bug triggered or caused by a plugin, such as this bug in pyresample, this should not cause trollflow2 to permanently stop processing.  There should be an option for a timeout, such that the user can configure that if a plugin hasn't completed after (for example) 5 minutes, trollflow2 gives up on this job and moves on to processing the next message.

**Actual results**

In operational version — see above.

The MCVE has the following output, then nothing, just 100% CPU usage.

```
[DEBUG: 2021-02-10 11:03:57 : trollsched.satpass] kwargs: {'instrument': 'avhrr/3'}
[DEBUG: 2021-02-10 11:03:57 : trollsched.satpass] instrument: avhrr/3
[DEBUG: 2021-02-10 11:03:57 : pyorbital.tlefile] Fetch TLE from the internet.
[DEBUG: 2021-02-10 11:03:58 : trollsched.boundary] Instrument = avhrr/3
^CTraceback (most recent call last):
  File "../debug-trollflow2.py", line 4, in <module>
    get_scene_coverage("Metop-C", datetime(2021, 2, 9, 14, 22, 3),
  File "/data/gholl/miniconda3/envs/py38b/lib/python3.8/site-packages/trollflow2/plugins/__init__.py", line 415, in get_scene_coverage
    return 100 * overpass.area_coverage(area_def)
  File "/data/gholl/miniconda3/envs/py38b/lib/python3.8/site-packages/trollsched/satpass.py", line 242, in area_coverage
    inter = self.boundary.contour_poly.intersection(area_boundary)
  File "/data/gholl/miniconda3/envs/py38b/lib/python3.8/site-packages/pyresample/spherical.py", line 495, in intersection
    return self._bool_oper(other, -1)
  File "/data/gholl/miniconda3/envs/py38b/lib/python3.8/site-packages/pyresample/spherical.py", line 476, in _bool_oper
    inter, edge2 = edge1.get_next_intersection(narcs2, inter)
  File "/data/gholl/miniconda3/envs/py38b/lib/python3.8/site-packages/pyresample/spherical.py", line 305, in get_next_intersection
    inter = self.intersection(arc)
  File "/data/gholl/miniconda3/envs/py38b/lib/python3.8/site-packages/pyresample/spherical.py", line 291, in intersection
    cd_ = c__.hdistance(d__)
  File "/data/gholl/miniconda3/envs/py38b/lib/python3.8/site-packages/pyresample/spherical.py", line 86, in hdistance
    return 2 * np.arcsin((np.sin((point.lat - self.lat) / 2.0) ** 2.0 +
KeyboardInterrupt
```

In the operational version, I have also used `pyrasite-shell` to connect to the stuck instance.  This reveals that trollflow2 is indeed stuck in the intersection calculation and that the ultimate cause is therefore https://github.com/pytroll/pyresample/issues/329:

```
>>> traceback.print_stack(sys._current_frames()[139653520578368])
  File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py38/bin/satpy_launcher.py", line 80, in <module>
    main()
  File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py38/bin/satpy_launcher.py", line 75, in main
    run(prod_list, topics=topics, test_message=test_message,
  File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py38/lib/python3.8/site-packages/trollflow2/launcher.py", line 152, in run
    proc.start()
  File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py38/lib/python3.8/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py38/lib/python3.8/multiprocessing/context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py38/lib/python3.8/multiprocessing/context.py", line 277, in _Popen
    return Popen(process_obj)
  File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py38/lib/python3.8/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py38/lib/python3.8/multiprocessing/popen_fork.py", line 75, in _launch
    code = process_obj._bootstrap(parent_sentinel=child_r)
  File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py38/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py38/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py38/lib/python3.8/site-packages/trollflow2/launcher.py", line 268, in process
    cwrk.pop('fun')(job, **cwrk)
  File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py38/lib/python3.8/site-packages/trollflow2/plugins/__init__.py", line 403, in covers
    cov = get_scene_coverage(platform_name, start_time, end_time,
  File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py38/lib/python3.8/site-packages/trollflow2/plugins/__init__.py", line 425, in get_scene_coverage
    return 100 * overpass.area_coverage(area_def)
  File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py38/lib/python3.8/site-packages/trollsched/satpass.py", line 242, in area_coverage
    inter = self.boundary.contour_poly.intersection(area_boundary)
  File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py38/lib/python3.8/site-packages/pyresample/spherical.py", line 494, in intersection
    return self._bool_oper(other, -1)
  File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py38/lib/python3.8/site-packages/pyresample/spherical.py", line 475, in _bool_oper
    inter, edge2 = edge1.get_next_intersection(narcs2, inter)
  File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py38/lib/python3.8/site-packages/pyresample/spherical.py", line 305, in get_next_intersection
    inter = self.intersection(arc)
  File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py38/lib/python3.8/site-packages/pyresample/spherical.py", line 284, in intersection
    for i in self.intersections(other_arc):
  File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py38/lib/python3.8/site-packages/pyresample/spherical.py", line 257, in intersections
    eb_ = other_arc.start.cross2cart(other_arc.end).normalize()
  File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py38/lib/python3.8/site-packages/pyresample/spherical.py", line 58, in cross2cart
    res = CCoordinate(np.array([-ad * c + be * f,
  File "/opt/pytroll/pytroll_inst/miniconda3/envs/pytroll-py38/lib/python3.8/site-packages/pyresample/spherical.py", line 112, in __init__
    self.cart = np.array(cart)
```

**Environment Info:**
 - OS: [e.g. OSX, Windows, Linux]
 - Trollflow2 Version: 0.8.1.dev7+g915b542
 - Satpy Version: 0.25.2.dev49+ge0f1fb8
 - Pyresample Version: 1.17.0+13.g270b5f0
 - supervisord version: 4.2.1

**Additional context**

I wonder if there could be a general timeout configured in message/job processing here:

https://github.com/pytroll/trollflow2/blob/6f9815ad141e14dfaa8fb7b2a53ec83b97c86e6f/trollflow2/launcher.py#L265-L270

although of course the pyresample bug needs to be fixed at all, in theory any plugin could get stuck for any or no reason at all, and it would be nice if trollflow2 could recover from that.

### PR DESCRIPTION
**Pull request by gerritholl**
Allow users to configure a timeout for each plugin/worker that is run.
If defined, this indicates the number in seconds trollflow2 will wait
for the job to complete.

 - [x] Closes #100 
 - [x] Tests added  
 - [x] Tests passed 
 - [x] Passes ``flake8 trollflow2`` 
 - [x] Fully documented 
 - [x] Add your name to `AUTHORS.md` if not there already

### CODE REVIEW

**Comment by gerritholl** [comment]
The test failure for `(3.9, ubuntu-latest, true)` (whatever `true` means here) does not appear related to this PR.

**Review by mraspaud — APPROVED** [review]
LGTM

File: doc/source/index.rst [hunk 1]
```diff
@@ -36,6 +36,11 @@ arguments are to be passed for the initialization of the class.
 If the callable has a ``stop`` method, it will be called without arguments at the
 end of each run (one scene).
 
+If the configuration has a ``timeout``, that will be used as the maximum time
+in seconds the plugin will be allowed to run.  If it has not completed within
+this number of seconds, the job will be considered to have failed.  This feature
+is only available on Unix.
```
Comment (pnuu) [inline doc/source/index.rst:42]: Maybe "POSIX compliant" or "Linux" depending which is closer, we don't formally support Unix :wink:

**Review by pnuu — APPROVED** [review]
LGTM. Just one nitpick on the documentation. And maybe an example in one of the configuration examples would be nice to have.

File: doc/source/index.rst (hunk 1 above)
Comment (gerritholl) [reply doc/source/index.rst:42]: I just took what it says in the Python documentation for [`signal.setitimer`](https://docs.python.org/3/library/signal.html#signal.setitimer).  What does trollflow2 formally support?  If trollflow2 is Linux-only I can remove the line entirely.  Has anyone tried to run trollflow2 on anything else than Linux?

**Comment by gerritholl** [comment]
Retry DeepCode

File: doc/source/index.rst (hunk 1 above)
Comment (mraspaud) [reply doc/source/index.rst:42]: trollflow is tested on linux and mac at the moment.

Comment (gerritholl) [reply doc/source/index.rst:42]: Linux and Mac are both flavours of Unix, aren't they?

Comment (gerritholl) [reply doc/source/index.rst:42]: Apparently my solution doesn't work on 3.7 for Mac, but does on 3.8 and 3.9.  Must I fix this, or can I skip this test for Mac on 3.7?  It's difficult for me to test on a Mac and I don't have access to one.  Or I can declare that this feature is supported only on Linux and skip the tests on Mac completely?

Comment (gerritholl) [reply doc/source/index.rst]: I've changed the text now to state it is only supported on Linux, and I've disabled the test for other platforms.

**Comment by gerritholl** [comment]
The CodeFactor failure is in a part of the code I did not touch.  I don't know why the (3.9, ubuntu-latest, true) test fails, but it doesn't appear to be in code I touched either.

**Comment by gerritholl** [comment]
The unstable test is failing if satpy is but pyresample isn't installed.  Is that really a situation we need to cover?

**Comment by gerritholl** [comment]
This test will pass if https://github.com/pytroll/trollflow2/pull/110 is merged first.

**Comment by mraspaud** [comment]
For some reason the tests of macos with python 3.7 are failing now...

**Comment by adybbroe** [comment]
LGTM. Thanks @gerritholl 
I would not bother about the CodeFactor issue for this PR!

**Review by adybbroe — APPROVED** [review]
LGTM. Just one question/request: Could you add an example using the timeout configuration in the example product list?

**Review by adybbroe — APPROVED** [review]
Many thanks, just want I wanted!

**Review by mraspaud — APPROVED** [review]
LGTM

**Pull request merged**